Fix security transparency issue

noahfalk opened the pull request:
Fixes #79749

[UPDATE]: The new fix is to remove the APTCA attribute and avoid running in partial trust.


[ORIGINAL fix that is no longer being applied]
This is a partial revert of #78729. That change inadvertently introduced a security transparency issue because the newly added AsSpan() call is security critical and can't be called in a security transparent scope.

stephentoub commented:
Why is AsSpan security critical, and why doesn't this impact every other netstandard2.0 library we have that uses AsSpan?

jkotas commented:
We have removed .NET Framework partial trust support long time ago. It is not possible to use any new language features for  .NET Framework partial trust. Details are in https://github.com/dotnet/runtime/issues/24122. 

Why does the test from https://github.com/dotnet/runtime/issues/79749 run in a partial trust sandbox?

noahfalk commented:
> Why does the test from https://github.com/dotnet/runtime/issues/79749 run in a partial trust sandbox?

Unknown. It isn't an intentional aspect of the tests for this assembly that it should run under partial trust. I assumed that partial trust was simply a test environment that was occasionally used and I had no desire for these tests to generate spurious failures if they were going to be subjected to that.

> why doesn't this impact every other netstandard2.0 library we have that uses AsSpan?

Perhaps because System.Diagnostics.DiagnosticSource has a lingering APTCA attribute and other libraries don't?
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Diagnostics.DiagnosticSource/src/AssemblyInfo.netfx.cs#L6

I am unaware of what scenario requires S.D.DS to maintain the APTCA attribute so I am fine to remove it. It sounds like that would align this assembly with general approach used by the rest of our libraries?

stephentoub commented:
> Perhaps because System.Diagnostics.DiagnosticSource has a lingering APTCA attribute and other libraries don't?

Got it.  Yeah, it appears to be the only one left.  (We also seem to have one test that sets it, but no other production assemblies.)

> so I am fine to remove it

That sounds like the right solution to me.

noahfalk committed:
Remove APTCA

None of our assemblies support partial trust usage any longer. Removing
the APTCA attribute from S.D.DS so that it follows the pattern.

Fixes #79749

jkotas commented:
> I am unaware of what scenario requires S.D.DS to maintain the APTCA attribute

FWIW, the APTCA attribute was introduced by https://github.com/dotnet/corefx/pull/17076 .

jkotas commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -116,7 +116,6 @@ System.Diagnostics.DiagnosticSource</PackageDescription>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
     <Compile Include="System\Diagnostics\HttpHandlerDiagnosticListener.cs" />
```
You should be also able delete `ALLOW_PARTIALLY_TRUSTED_CALLERS` from this file and all code under this ifdef.

noahfalk replied:
Got rid of the rest of it.

noahfalk committed:
Remove more code for partial trust

jkotas commented:
Could you please also delete `ALLOW_PARTIALLY_TRUSTED_CALLERS` from `src\libraries\Common\src\System\HexConverter.cs`?

LGTM otherwise.

noahfalk committed:
More partial trust removal

noahfalk commented:
Sure, got HexConverter.cs as well.

stephentoub commented:
It'd still be interesting to know why the tests are running in partial trust. @ViktorHofer, do you know?

jkotas commented:
I do not think that the tests are running in partial trust. 

.NET Framework validates consistency of the SecurityCritical/SecurityTransparent attributes in APTCA assemblies even when the code is running in full trust. It explains why the test is failing.

stephentoub commented:
> even when the code is running in full trust

Interesting, I didn't know that. Thanks.
